### PR TITLE
VL53L1X Region of Interest & Distance Mode Setting

### PR DESCRIPTION
**Pull request by dirksavage88**
**Describe problem solved by this pull request**
Distance sensor driver for the VL53L1X has region of interest (ROI) that can be programmed, but the set of functions were not implemented in the VL53L1X ul driver integrated into PX4. The ROI settings allow better resolution when changing the ranging area within the field of view. Distance is decreased but resolution is increased allowing the sensor to pick up smaller objects at sub-1 meter ranges.

**Describe your solution**
Utilized the ST vl53l1x ultra lite driver manual to implement two functions that allow a programmable region of interest, also created a distance mode class data member that allows easier setting, this defaults to a short distance mode for better resolution. These two settings lay the ground work for using multiple simultaneous zone detection with multiple distance sensors (2d lidar)

**Describe possible alternatives**
Utilize code from a larger driver that allows Lidar sensing with multiple sensors. However, this has larger code footprint.

**Test data / coverage**
Tested on bench setting with real hardware ITL. Plot from flight review:
bokeh_plot
 attached

### CODE REVIEW

**Comment by dagar** [comment]
Interesting, I was considering directly carrying the (big ugly) ST driver to make more of these features accessible.

File: src/modules/mavlink/mavlink_receiver.cpp [hunk 1]
```diff
@@ -765,6 +765,17 @@ MavlinkReceiver::handle_message_distance_sensor(mavlink_message_t *msg)
 	// quality value. Also it comes normalised between 1 and 100 while the uORB
 	// signal quality is normalised between 0 and 100.
 	ds.signal_quality = dist_sensor.signal_quality == 0 ? -1 : 100 * (dist_sensor.signal_quality - 1) / 99;
+        qt = const_cast<float*>(ds.q);
+	//mavlink clients
+        mavlink_msg_distance_sensor_send(MAVLINK_COMM_0, ds.timestamp, ds.min_distance, ds.max_distance, ds.current_distance, ds.type, ds.device_id, ds.orientation, ds.variance, ds.h_fov, ds.v_fov, qt, ds.signal_quality);
```
Comment (dagar) [inline src/modules/mavlink/mavlink_receiver.cpp:770]: Crossing mavlink channels (owned by separate mavlink instances) doesn't look quite right, could you explain what you're trying to accomplish here?

File: src/drivers/distance_sensor/vl53l1x/vl53l1x.cpp [hunk 2]
```diff
@@ -222,35 +232,55 @@ void VL53L1X::RunImpl()
 {
 	uint8_t dataReady = 0;
 
+	uint8_t roiCenter[] = {247, 215, 183, 167, 151};
```
Comment (dagar) [inline src/drivers/distance_sensor/vl53l1x/vl53l1x.cpp:235]: Comment explain these magic numbers?

Is it desirable to configure a specific ROI for different usage?

File: src/drivers/distance_sensor/vl53l1x/vl53l1x.cpp [hunk 3]
```diff
@@ -166,16 +176,16 @@ VL53L1X::~VL53L1X()
...
 
 	perf_begin(_sample_perf);
 
 	const hrt_abstime timestamp_sample = hrt_absolute_time();
 
 	ret = VL53L1X_GetRangeStatus(&rangeStatus);
-
-	if ((ret != PX4_OK) | (rangeStatus != PX4_OK)) {
+	
+	if ((ret != PX4_OK) | (rangeStatus == 13)) {
```
Comment (dagar) [inline src/drivers/distance_sensor/vl53l1x/vl53l1x.cpp:188]: Is there a range status define you can use instead of the magic number (13)?

File: src/modules/mavlink/mavlink_receiver.cpp (hunk 1 above)
Comment (dirksavage88) [reply src/modules/mavlink/mavlink_receiver.cpp:770]: I'll be honest, I was troubleshooting mavlink messages issues. Turns out the sensor itself didn't like 5 volts and would refuse to work when connected to a LiPO, but would work perfectly fine in USB. Thought it was a telemetry vs usb issue. Can remove, unnecessary.

File: src/drivers/distance_sensor/vl53l1x/vl53l1x.cpp (hunk 2 above)
Comment (dirksavage88) [reply src/drivers/distance_sensor/vl53l1x/vl53l1x.cpp:235]: logic to these numbers can be found in vl53l1x multiple detection zone pdf.

File: src/drivers/distance_sensor/vl53l1x/vl53l1x.cpp (hunk 3 above)
Comment (dirksavage88) [reply src/drivers/distance_sensor/vl53l1x/vl53l1x.cpp:188]: this is a range status that indicates ROI is 'out of bounds', also referenced in the ST multiple zone manaul

File: src/modules/mavlink/mavlink_receiver.cpp (hunk 1 above)
Comment (LorenzMeier) [reply src/modules/mavlink/mavlink_receiver.cpp:770]: Please remove yes.

File: src/drivers/distance_sensor/vl53l1x/vl53l1x.cpp (hunk 3 above)
Comment (LorenzMeier) [reply src/drivers/distance_sensor/vl53l1x/vl53l1x.cpp:188]: Please create const values for each of those, e.g. RANGE_STATUS_OUT_OF_BOUNDS for this one and then use it.

File: src/drivers/distance_sensor/vl53l1x/vl53l1x.cpp [hunk 4]
```diff
@@ -144,11 +144,21 @@ VL53L1X::VL53L1X(I2CSPIBusOption bus_option, const int bus, const uint8_t rotati
...
+{       
+	//Set distance mode (1 for ~2m ranging, 2 for ~4m ranging
+	distance_mode = 1;
+
+	// VL53L1X typical range 0-4 meters with 27 degree field of view
 	_px4_rangefinder.set_min_distance(0.f);
-	_px4_rangefinder.set_max_distance(2.f);
-	_px4_rangefinder.set_fov(math::radians(25.f));
+
+	if (distance_mode == 1) {
```
Comment (LorenzMeier) [inline src/drivers/distance_sensor/vl53l1x/vl53l1x.cpp:154]: Please use defines instead of magic numbers.

File: src/drivers/distance_sensor/vl53l1x/vl53l1x.cpp [hunk 5]
```diff
@@ -166,16 +176,16 @@ VL53L1X::~VL53L1X()
 int VL53L1X::collect()
 {
 	uint8_t ret = 0;
-	uint8_t rangeStatus;
+	uint8_t rangeStatus = 0;
```
Comment (LorenzMeier) [inline src/drivers/distance_sensor/vl53l1x/vl53l1x.cpp:179]: Needs a define here as well.

File: src/drivers/distance_sensor/vl53l1x/vl53l1x.cpp [hunk 6]
```diff
@@ -191,7 +201,7 @@ int VL53L1X::collect()
 	}
 
 	perf_end(_sample_perf);
-
+
```
Comment (LorenzMeier) [inline src/drivers/distance_sensor/vl53l1x/vl53l1x.cpp:204]: Please remove this whitespace change (run `make format`)

File: src/drivers/distance_sensor/vl53l1x/vl53l1x.cpp (hunk 2 above)
Comment (LorenzMeier) [reply src/drivers/distance_sensor/vl53l1x/vl53l1x.cpp:235]: Please use defines to encode them.

File: src/drivers/distance_sensor/vl53l1x/vl53l1x.hpp [hunk 7]
```diff
@@ -123,13 +123,16 @@ class VL53L1X : public device::I2C, public I2CSPIDriver<VL53L1X>
 	void stop();
 
 	virtual int init() override;
-
+
```
Comment (LorenzMeier) [inline src/drivers/distance_sensor/vl53l1x/vl53l1x.hpp:126]: Unnecessary whitespace change.

File: boards/nxp/fmuk66-v3/default.cmake [hunk 8]
```diff
@@ -22,7 +22,7 @@ px4_add_board(
 		camera_trigger
 		differential_pressure # all available differential pressure drivers
 		distance_sensor # all available distance sensor drivers
-		distance_sensor/srf05 # Specific driver
+		distance_sensor/srf05 # Specific driver
```
Comment (LorenzMeier) [inline boards/nxp/fmuk66-v3/default.cmake:25]: Unnecessary whitespace change

File: src/drivers/distance_sensor/vl53l1x/vl53l1x.cpp [hunk 9]
```diff
@@ -144,11 +144,21 @@ VL53L1X::VL53L1X(I2CSPIBusOption bus_option, const int bus, const uint8_t rotati
 	I2C(DRV_DIST_DEVTYPE_VL53L1X, MODULE_NAME, bus, address, bus_frequency),
 	I2CSPIDriver(MODULE_NAME, px4::device_bus_to_wq(get_device_id()), bus_option, bus),
 	_px4_rangefinder(get_device_id(), rotation)
-{
-	// VL53L1X typical range 0-2 meters with 25 degree field of view
+{       
+	//Set distance mode (1 for ~2m ranging, 2 for ~4m ranging
```
Comment (LorenzMeier) [inline src/drivers/distance_sensor/vl53l1x/vl53l1x.cpp:148]: Please use one space after `//` for every comment

File: src/drivers/distance_sensor/vl53l1x/vl53l1x.cpp [hunk 10]
```diff
@@ -222,35 +232,55 @@ void VL53L1X::RunImpl()
...
 	uint8_t dataReady = 0;
 
+	uint8_t roiCenter[] = {247, 215, 183, 167, 151}; 
+        static uint8_t zone = 0;
+
 	VL53L1X_CheckForDataReady(&dataReady);
 
 	if (dataReady == 1) {
-		collect();
+                collect();
```
Comment (LorenzMeier) [inline src/drivers/distance_sensor/vl53l1x/vl53l1x.cpp:241]: Whitespace / indentation change that should be removed (wrong indentation and it adds noise to the diff)

**Comment by mrpollo** [comment]
@dirksavage88 thanks for putting in the work to push this upstream, we would love to get this in soon, can you please check the CI errors and comments from the reviewers above? once you are ready for a new review please ping me

**Comment by dirksavage88** [comment]
> @dirksavage88 thanks for putting in the work to push this upstream, we would love to get this in soon, can you please check the CI errors and comments from the reviewers above? once you are ready for a new review please ping me

Hello @mrpollo , 

I have incorporated the changes in the latest commit, to include format changes. **I have successfully built for my specific hardware target but not tested (SITL, HITL, etc).**

**Comment by mrpollo** [comment]
Thanks @dirksavage88 

Looks like the CI timeout is non-related to your changeset, I will ask for a final review from the team before merging.

File: boards/nxp/fmuk66-v3/default.cmake [hunk 11]
```diff
@@ -22,7 +22,7 @@ px4_add_board(
 		camera_trigger
 		differential_pressure # all available differential pressure drivers
 		distance_sensor # all available distance sensor drivers
-		distance_sensor/srf05 # Specific driver
+        distance_sensor/srf05 # Specific driver
```
Comment (dagar) [inline boards/nxp/fmuk66-v3/default.cmake:25]:
```suggestion
		distance_sensor/srf05 # Specific driver
```

File: src/drivers/distance_sensor/vl53l1x/vl53l1x.cpp [hunk 12]
```diff
@@ -222,35 +241,56 @@ void VL53L1X::RunImpl()
...
 	// Schedule the first cycle.
 	ScheduleNow();
+
 }
 
 int VL53L1X::init()
 {
 	int ret = PX4_OK;
-
+	uint8_t x, y;
```
Comment (dagar) [inline src/drivers/distance_sensor/vl53l1x/vl53l1x.cpp:278]: This isn't C89, you can declare these where you first use them below.

Comment (dirksavage88) [reply src/drivers/distance_sensor/vl53l1x/vl53l1x.cpp]: done

File: boards/nxp/fmuk66-v3/default.cmake (hunk 11 above)
Comment (dirksavage88) [reply boards/nxp/fmuk66-v3/default.cmake]: fixed

**Comment by dirksavage88** [comment]
@mrpollo Wasn't sure if the CI failures are related, but the SITL tests /build (tailsitter) and MAVROS mission tests / build rover, are those fails that can be remedied?

**Comment by mrpollo** [comment]
@dirksavage88 I don't think either of those failed tests has anything to do with your PR, I think we should proceed with a merge as soon as a reviewer with commit access has time to get this in.

I'm tracking the CI failures here in case anyone is interested in following up #16859 and #16745

**Comment by igalloway** [comment]
Thanks @dirksavage88 
@LorenzMeier @dagar @mrpollo  FYI - This is a HoverGames2 contestant
https://www.hackster.io/savagecabbage/smart-scan-delivery-drone-ed5c54

File: boards/nxp/fmuk66-v3/default.cmake [hunk 13]
```diff
@@ -22,7 +22,7 @@ px4_add_board(
 		camera_trigger
 		differential_pressure # all available differential pressure drivers
 		distance_sensor # all available distance sensor drivers
-		distance_sensor/srf05 # Specific driver
+                distance_sensor/srf05 # Specific driver
```
Comment (dagar) [inline boards/nxp/fmuk66-v3/default.cmake:25]:
```suggestion
		distance_sensor/srf05 # Specific driver
```

File: src/drivers/distance_sensor/vl53l1x/vl53l1x.cpp [hunk 14]
```diff
@@ -704,6 +761,7 @@ extern "C" __EXPORT int vl53l1x_main(int argc, char *argv[])
 
 	if (!strcmp(verb, "start")) {
 		return ThisDriver::module_start(cli, iterator);
+
```
Comment (dagar) [inline src/drivers/distance_sensor/vl53l1x/vl53l1x.cpp:764]:
```suggestion
```

File: src/drivers/distance_sensor/vl53l1x/vl53l1x.hpp [hunk 15]
```diff
@@ -156,9 +159,11 @@ class VL53L1X : public device::I2C, public I2CSPIDriver<VL53L1X>
...
 	int8_t VL53L1X_ConfigBig(uint16_t DM, uint16_t TimingBudgetInMs);
 	int8_t VL53L1X_SetInterMeasurementInMs(uint32_t InterMeasurementInMs);
-
+	int8_t VL53L1X_SetROICenter(uint8_t data);
+	int8_t VL53L1X_SetROI(uint16_t x, uint16_t y);
 	PX4Rangefinder    _px4_rangefinder;
 
 	perf_counter_t _comms_errors{perf_alloc(PC_COUNT, MODULE_NAME": com_err")};
 	perf_counter_t _sample_perf{perf_alloc(PC_ELAPSED, MODULE_NAME": read")};
+
```
Comment (dagar) [inline src/drivers/distance_sensor/vl53l1x/vl53l1x.hpp:168]:
```suggestion
```

File: src/drivers/distance_sensor/vl53l1x/vl53l1x.cpp [hunk 16]
```diff
@@ -222,35 +241,55 @@ void VL53L1X::RunImpl()
...
+	}
+
 }
 
 void VL53L1X::start()
 {
+
 	// Schedule the first cycle.
 	ScheduleNow();
+
```
Comment (dagar) [inline src/drivers/distance_sensor/vl53l1x/vl53l1x.cpp:272]:
```suggestion
```

File: src/drivers/distance_sensor/vl53l1x/vl53l1x.cpp [hunk 17]
```diff
@@ -222,35 +241,55 @@ void VL53L1X::RunImpl()
...
+	// reset counter check
+	if (zone > sizeof(roiCenter) - 1) {
+		zone = 0;
+	}
+
 }
 
 void VL53L1X::start()
 {
+
```
Comment (dagar) [inline src/drivers/distance_sensor/vl53l1x/vl53l1x.cpp:269]:
```suggestion
```

File: src/drivers/distance_sensor/vl53l1x/vl53l1x.hpp [hunk 18]
```diff
@@ -130,6 +130,9 @@ class VL53L1X : public device::I2C, public I2CSPIDriver<VL53L1X>
 	 */
 	void RunImpl();
 
+	// Distance mode member variable
+	uint16_t distance_mode;
```
Comment (dagar) [inline src/drivers/distance_sensor/vl53l1x/vl53l1x.hpp:134]:
```suggestion
	uint16_t distance_mode{VL53L1X_SHORT_RANGE};
```

File: src/drivers/distance_sensor/vl53l1x/vl53l1x.cpp [hunk 19]
```diff
@@ -145,10 +154,20 @@ VL53L1X::VL53L1X(I2CSPIBusOption bus_option, const int bus, const uint8_t rotati
 	I2CSPIDriver(MODULE_NAME, px4::device_bus_to_wq(get_device_id()), bus_option, bus),
 	_px4_rangefinder(get_device_id(), rotation)
 {
-	// VL53L1X typical range 0-2 meters with 25 degree field of view
+	// Set distance mode (1 for ~2m ranging, 2 for ~4m ranging
+	distance_mode = VL53L1X_SHORT_RANGE;
```
Comment (dagar) [inline src/drivers/distance_sensor/vl53l1x/vl53l1x.cpp:158]:
```suggestion
```

File: src/drivers/distance_sensor/vl53l1x/vl53l1x.cpp [hunk 20]
```diff
@@ -145,10 +154,20 @@ VL53L1X::VL53L1X(I2CSPIBusOption bus_option, const int bus, const uint8_t rotati
 	I2CSPIDriver(MODULE_NAME, px4::device_bus_to_wq(get_device_id()), bus_option, bus),
 	_px4_rangefinder(get_device_id(), rotation)
 {
-	// VL53L1X typical range 0-2 meters with 25 degree field of view
+	// Set distance mode (1 for ~2m ranging, 2 for ~4m ranging
```
Comment (dagar) [inline src/drivers/distance_sensor/vl53l1x/vl53l1x.cpp:157]:
```suggestion
```

File: src/drivers/distance_sensor/vl53l1x/vl53l1x.cpp [hunk 21]
```diff
@@ -145,10 +154,20 @@ VL53L1X::VL53L1X(I2CSPIBusOption bus_option, const int bus, const uint8_t rotati
...
+	distance_mode = VL53L1X_SHORT_RANGE;
+
```
Comment (dagar) [inline src/drivers/distance_sensor/vl53l1x/vl53l1x.cpp:159]:
```suggestion
```

**Comment by dagar** [comment]
Can you get rid of the submodule changes?

**Comment by dirksavage88** [comment]
Yes

**Comment by dirksavage88** [comment]
> Can you get rid of the submodule changes?

revert back to d5401fa?
